lib/elixir/lib/integer.ex

```diff
@@ -240,45 +240,30 @@ defmodule Integer do
   end
 
   def parse(<<?-, rest::binary>>, base) do
-    case parse_digits(rest, base) do
-      {acc, bin} ->
-        {-acc, bin}
-      :error ->
-        :error
-    end
-  end
-
-  def parse(<<?+, rest::binary>>, base) do
-    parse_digits(rest, base)
+    with {int, rem} <- parse_digits(rest, base), do: {-int, rem}
   end
-
-  def parse(<<rest::binary>>, base) do
-    parse_digits(rest, base)
-  end
-
-  digits = [{?0..?9, -?0}, {?A..?Z, 10 - ?A}, {?a..?z, 10 - ?a}]
-
-  for {chars, diff} <- digits, char <- chars do
-    defp parse_digits(<<unquote(char), rest::binary>>, base)
-         when base > unquote(char + diff) do
-      parse_digits(rest, base, unquote(char + diff))
+  def parse(<<?+, rest::binary>>, base), do: parse_digits(rest, base)
+  def parse(<<rest::binary>>, base), do: parse_digits(rest, base)
+
+  defp parse_digits(<<rest::binary>>, base) do
+    case count_digits(rest, base, 0) do
+      0 -> :error
+      n ->
+        {digits, rem} = :erlang.split_binary(rest, n)
```

Comment:
> We can avoid the - appearing string bug because the count argument is matching on 0, so must appear at the beginning of the string.

I understand now, that makes sense.

I've moved the +/- handling down to `count_digits` in [this branch](https://github.com/balexand/elixir/compare/balexand:integer_parse...balexand:count_digits). It has about the same performance, but I think the code is uglier.

> `rest` is used twice

And it's still used twice. The function where `rest` is used twice isn't doing anything related to +/- handling.